fundraiser/src/main.sw

```diff
@@ -0,0 +1,502 @@
+        let mut amount = amount; // workaround until `mut` is able to be set as a param
+
+        // If the user is attempting to unpledge more than they have pledged then reset the amount
+        // they are withdrawing to the maximum that they have pledged to this campaign
+        if pledge.amount < amount {
+            amount = pledge.amount;
+        }
+
+        // Update the amount that they have pledged
+        pledge.amount = pledge.amount - amount;
+
+        // Lower the campaign total pledge by the amount the user has unpledged
+        campaign_info.total_pledge = campaign_info.total_pledge - amount;
+
+        // Update the state of their pledge with the new version
+        storage.pledge_history.insert((user, pledge_history_index), pledge);
+
+        // Update the campaign state with the updated version as well
+        storage.campaign_info.insert(id, campaign_info);
+
+        // Update the asset amount to track the removal of the amount
+        let mut asset_info = storage.asset_info.get(campaign_info.asset);
+        asset_info.amount = asset_info.amount - amount;
+
+        // Update asset state
+        storage.asset_info.insert(campaign_info.asset, asset_info);
+
+        // Transfer back the amount the user has unpledged
+        transfer(amount, campaign_info.asset, user);
+
+        // We have updated the state of a campaign therefore we must log it
+        log(UnpledgedEvent {
+            amount, id, user
+        });
+    }
+
+    /// Returns the total number of campaigns that have been created by all users
+    #[storage(read)]fn total_campaigns() -> u64 {
+        storage.total_campaigns
+    }
+
+    /// Returns information about the specified campaign
+    ///
+    /// # Arguments
+    ///
+    /// * `id` - Unique campaign identifier which is a number from the storage.total_campaigns range
+    ///
+    /// # Reverts
+    ///
+    /// * When the `id` is either 0 or greater than the total number of campaigns created
+    #[storage(read)]fn campaign_info(id: u64) -> CampaignInfo {
+        // User cannot interact with a non-existent campaign
+        validate_id(id, storage.total_campaigns);
+        storage.campaign_info.get(id)
+    }
+
+    /// Returns the number of campaigns that the user has created
+    ///
+    /// # Reverts
+    ///
+    /// * When an AuthError is generated
+    #[storage(read)]fn user_campaign_count() -> u64 {
+        storage.user_campaign_count.get(sender_identity())
+    }
+
+    /// Returns information about the specified campaign for the campaign author
+    ///
+    /// # Arguments
+    ///
+    /// * `id` - Unique identifier which is a number starting from 1...storage.user_campaign_count
+    ///
+    /// # Reverts
+    ///
+    /// * When the `id` is either 0 or greater than the total number of campaigns created by the author
+    /// * When an AuthError is generated
+    #[storage(read)]fn campaign(id: u64) -> Campaign {
```

Comment:
I don't think we need to change this now, but in general, the approach of using `msg_sender` for this function (rather than accepting a second param like `user: Identity`) only gives a small amount of pseudo-privacy, while preventing possible future integrations (i.e: A third-party contract that interacts with the fundraiser contract on behalf of a user). Passing a second param here would enable composability in the greater ecosystem (which is currently non-existant, so could be a future improvement to make).